### PR TITLE
check that layered nodes use CoreOS

### PR DESCRIPTION
**Pull request by mkenigs**
Layered updates are not possible on non-CoreOS nodes

### CODE REVIEW

File: pkg/daemon/daemon.go [hunk 1]
```diff
@@ -1470,9 +1470,16 @@ func (dn *Daemon) removeUpdateInProgressTaint(ctx context.Context) error {
...
 }
 
-func (dn *Daemon) isLayeredNode() bool {
+func (dn *Daemon) isLayeredNode() *CoreOSDaemon {
 	_, ok := dn.node.Annotations[constants.DesiredImageConfigAnnotationKey]
-	return ok
+	if ok {
+		if dn.os.IsCoreOSVariant() {
+			return &CoreOSDaemon{dn}
+		} else {
```
Comment (mkenigs) [inline pkg/daemon/daemon.go:1478]: Should we degrade or something here?

Comment (kikisdeliveryservice) [reply pkg/daemon/daemon.go:1478]: I don't think degrading is the right choice bc there isn't technically a problem with the cluster.

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -1470,9 +1470,16 @@ func (dn *Daemon) removeUpdateInProgressTaint(ctx context.Context) error {
...
+			glog.Info("refusing to treat a non-CoreOS node as layered. You should consider using CoreOS though, it's pretty cool")
```
Comment (kikisdeliveryservice) [inline pkg/daemon/daemon.go:1479]: also while i agree, probably fix this message so it never goes into prod ;)

File: pkg/daemon/daemon.go (hunk 1 above)
Comment (mkenigs) [reply pkg/daemon/daemon.go:1478]: Okay so go ahead and perform a normal update, even though the node is labeled layered?

File: pkg/daemon/daemon.go (hunk 2 above)
Comment (mkenigs) [reply pkg/daemon/daemon.go:1479]: Haha okay will do

File: pkg/daemon/daemon.go [hunk 3]
```diff
@@ -1470,9 +1470,16 @@ func (dn *Daemon) removeUpdateInProgressTaint(ctx context.Context) error {
 	})
 }
 
-func (dn *Daemon) isLayeredNode() bool {
+func (dn *Daemon) isLayeredNode() *CoreOSDaemon {
 	_, ok := dn.node.Annotations[constants.DesiredImageConfigAnnotationKey]
```
Comment (cgwalters) [inline pkg/daemon/daemon.go:1474]: I think actually the controller should not allow this situation to be set up.  It also makes sense to have a last-ditch sanity check on the daemon side.

I wouldn't conflate checking "has a desired image config annotation" with "is coreos" check though, they seem like separate things.

Comment (mkenigs) [reply pkg/daemon/daemon.go:1474]: Would `isLayeredNode -> shouldUpdateLayered` clarify that? Or is that squashing checks together unclearly

File: pkg/daemon/daemon.go (hunk 1 above)
Comment (kikisdeliveryservice) [reply pkg/daemon/daemon.go:1478]: I agree with Colin below that a non-CoreOS node just shouldn't hit this via the Controller

Comment (mkenigs) [reply pkg/daemon/daemon.go:1478]: Oh okay yeah that'd be better. I'll look into fixing that on the controller side...

File: pkg/daemon/daemon.go (hunk 2 above)
Comment (mkenigs) [reply pkg/daemon/daemon.go]: Removed

File: pkg/daemon/daemon.go [hunk 4]
```diff
@@ -1470,9 +1470,17 @@ func (dn *Daemon) removeUpdateInProgressTaint(ctx context.Context) error {
...
 
-func (dn *Daemon) isLayeredNode() bool {
+func (dn *Daemon) shouldUpdateLayered() *CoreOSDaemon {
 	_, ok := dn.node.Annotations[constants.DesiredImageConfigAnnotationKey]
-	return ok
+	if ok {
+		// sanity check that we're on CoreOS. The controller checks if a node runs CoreOS before setting desired image
+		if dn.os.IsCoreOSVariant() {
+			return &CoreOSDaemon{dn}
+		} else {
```
Comment (kikisdeliveryservice) [inline pkg/daemon/daemon.go:1479]:
```suggestion
	
```
don't need the else you can just log and return nil

Comment (mkenigs) [reply pkg/daemon/daemon.go]: Fixed

File: pkg/controller/node/node_controller.go [hunk 5]
```diff
@@ -976,22 +980,32 @@ func (ctrl *Controller) updateCandidateMachines(pool *mcfgv1.MachineConfigPool,
...
-		// as our "done" signal
-		if isImagePool {
-			if targetImageMatchesConfig {
-				ctrl.logPool(pool, "Setting node %s target image to %s", node.Name, targetImage)
-				if err := ctrl.setNodeAnnotation(node.Name, daemonconsts.DesiredImageConfigAnnotationKey, targetImage); err != nil {
-					return goerrs.Wrapf(err, "setting desired config for node %s", node.Name)
+		// double check that this node has CoreOS and will support layering
+		osID, ok := node.Labels[osIDLabel]
+		if ok {
+			if daemon.IsProbablyCoreOSVariant(osID) {
```
Comment (mkenigs) [inline pkg/controller/node/node_controller.go:986]: Is it okay for me to use daemon code in the node_controller?

Comment (cgwalters) [reply pkg/controller/node/node_controller.go:986]: I don't see the problem with that, though I guess it looks like `ctrlcommon` is a bit cleaner.

**Comment by mkenigs** [comment]
/hold
Would like to hear what @jkyros thinks of controller changes

File: pkg/controller/node/node_controller.go [hunk 6]
```diff
@@ -976,22 +980,32 @@ func (ctrl *Controller) updateCandidateMachines(pool *mcfgv1.MachineConfigPool,
...
+							return goerrs.Wrapf(err, "setting desired config for node %s", node.Name)
+						}
+					} else {
+						glog.Infof("Image %s matched %s not %s. Proper image may not have rendered yet", targetImage, equivalentTo, targetConfig)
+					}
+
 				}
 			} else {
-				glog.Infof("Image %s matched %s not %s. Proper image may not have rendered yet", targetImage, equivalentTo, targetConfig)
+				glog.Infof("falling back to legacy update for non-CoreOS node %s; node has OS ID %s", node.Name, osID)
```
Comment (jkyros) [inline pkg/controller/node/node_controller.go:1005]: We've kind of been implicitly saying that "layering" is pool-granular, and it has to be once our content `spec.Configuration` is an Image. So while this works today with the annotations to where we can go either/or, this won't work anymore once we set `spec.Configuration` to an image, and I don't know that we want it to? 

If we don't like that it's pool-granular, we probably need to worry about those consequences now before we get too far. 

(If we're not okay with the trades and we do want "heterogeneous pools" then our pool probably has multiple payloads available that are vaguely equivalent and get assigned to nodes based on what kind they are, and we should probably mess with our CRD to accommodate that)

File: pkg/daemon/daemon.go (hunk 3 above)
Comment (jkyros) [reply pkg/daemon/daemon.go:1474]: Yeah,  `isLayeredNode` is more "has this node been assigned a desired image configuration", which in our human description makes it a "layered node" but it's more a quality of the update being requested rather than a quality of the node. 

If we had `canUpdateLayered()` in the daemon and one of the sanity checks in there was "is coreos" I think that would make sense? 


As an aside, should we just call it an "Image update"? Then we have "image update" vs "machineconfig update" and it's just referring to the type of the object in `spec.Configuration` ? (Unless you think we want a new CRD for holding whatever a layered update looks like, kinda like machineconfig... :stuck_out_tongue: ) 

(And yeah, I'm well aware I had a hand in naming all this stuff how it is :smile: )

File: pkg/daemon/daemon.go [hunk 7]
```diff
@@ -1505,8 +1512,8 @@ func (dn *Daemon) triggerUpdateWithMachineConfig(currentConfig, desiredConfig *m
 	dn.stopConfigDriftMonitor()
 
 	// Hack in our layered node workflow for pools labeled as "layered"
-	if dn.isLayeredNode() {
-		return dn.experimentalUpdateLayeredConfig()
+	if coreOSDaemon := dn.shouldUpdateLayered(); coreOSDaemon != nil {
```
Comment (jkyros) [inline pkg/daemon/daemon.go:1515]: I feel like you're kind of getting at a thing with CoreOSDaemon where we have different "types"/interfaces of daemons that determine their "features" that we could check.

**Review by jkyros — COMMENTED** [review]
As things are set up now with the "halfway route" that's already merged to layering: 
- We get to a "desiredConfig" level via an image. 
- The image is assigned to the pool via annotations IN ADDITION to the spec.Configuration

I did that (instead of applying the image as spec.Config) for expedience and because: 
- I didn't immediately notice that pool.Spec.Configuration was an ObjectRef since we don't treat it like one and 
- It didn't hard break anything -- things that didn't care about the image could ignore it

And you've seen the ramifications of that in: 
- render_controller where I have to annotate the pool with the proper image 
- node_controller where I have to set the additional desiredImage annotations 
- daemon where you have to set "double status" to make both the image and the existing machineConfig annotations happy

So while it does work, and was a reasonable half-step, I don't think that's what we want based on our designs that want `spec.Configuration` to be the actual image.  

**TL;DR we need to either:** 
1. Accept our half-step "annotation dance" and develop around that for 4.11 knowing we will throw some % of it away (not my preference) 
2. Think through the ramifications of `spec.Configuration` being an Image and develop around that instead.

File: pkg/daemon/daemon.go (hunk 3 above)
Comment (mkenigs) [reply pkg/daemon/daemon.go:1474]: Yeah I think image update makes sense! Machine config vs image is a lot more clear than what I've been saying (legacy vs layered)

**Comment by jkyros** [comment]
None of this hurts anything, and @mkenigs said he needed it for kernel args. We're going to have to tinker with these bits again for the build controller anyway, I don't want to block until then.

File: pkg/daemon/daemon.go (hunk 3 above)
Comment (cgwalters) [reply pkg/daemon/daemon.go:1478]: > As an aside, should we just call it an "Image update"? Then we have "image update" vs "machineconfig update" and 

Yes!  SGTM

File: pkg/controller/node/node_controller.go (hunk 6 above)
Comment (cgwalters) [reply pkg/controller/node/node_controller.go]: I don't think we need to care about pool-granular right now *except* that we will likely need at some point to be able to handle in-place update transitions to images.
